log: miscellaneous improvements.

knz opened the pull request:
Fixes  #11654.

Highlights:
- this checks that an empty `--log-dir` effectively disables logging to files
- new flag: `--log-file-max-size` (default: 10MiB)
- new flag: `--log-dir-max-size` (default: 100MiB)
- new flag: `--log-file-verbosity` to limit what goes to logs (default: all)
- `--alsologtostderr` is removed; a stderr severity threshold can now be set simply with `--logtostderr`.

knz committed:
cli: move bytesValue to humanizeutil.BytesValue.

knz committed:
log: allow the user to configure the log max sizes on the command line.

This patch introduces the two command-line parameters
`-log-file-max-size` and `-log-dir-max-size` to set the maximum size
of one log file and all combined log files, respectively.

Note that there is a small race condition: the log file GC process is
spawned before command-line argument handling and will thus for a
short time upon process initialization use the default settings.  This
may be an issue if running the process using a log directory with a
large amount of logs (more than 100MiB, the default max size) and if
the user intends to keep these. This will be addressed in a later patch.

knz committed:
log: remove dead code.

knz committed:
log: move the definition of -no-redirect-stderr where it belongs.

knz committed:
log: allow the user to filter which messages go to the log file.

This patch introduces a new command-line parameter
`-log-file-verbosity` which determines the minimum severity level for
messages written to the log file.

knz committed:
cli: add a test that checks --log-dir.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14841)
<!-- Reviewable:end -->

tamird commented:
Reviewed 2 of 2 files at r1, 7 of 7 files at r2, 1 of 1 files at r3, 3 of 3 files at r4, 5 of 5 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/util/humanizeutil/humanize.go, line 84 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXYkep5aVCeuS4XdOH:-KhXYkeqYnE_Tqw0JyYk:b-xkh1st) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/humanizeutil/humanize.go#L84)):*
> ```Go
> 		return err
> 	}
> 	atomic.StoreInt64(b.val, v)
> ```

this doesn't seem to have anything to do with what the commit message claims. why was it necessary? what's the race?

---

*[pkg/util/log/clog.go, line 902 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXYzPy2saFi2zv-AIj:-KhXYzPz2EZDVgHBhHWr:b1uo04z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/log/clog.go#L902)):*
> ```Go
> 
> func (sb *syncBuffer) Write(p []byte) (n int, err error) {
> 	if sb.nbytes+int64(len(p)) >= atomic.LoadInt64(&LogFileMaxSize) {
> ```

ditto - what's the race? here and below.

---

*[pkg/util/log/clog.go, line 89 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXbGz9mlDlvO4gazof:-KhXbGz9mlDlvO4gazog:bbpq0b5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2a63d5851900063acfa188d957c7d6315323ade/pkg/util/log/clog.go#L89)):*
> ```Go
> 		threshold = Severity(v)
> 	}
> 	logging.stderrThreshold.set(threshold)
> ```

was this just...completely wrong before?

why does it need to be atomic now?

---

*[pkg/util/log/clog.go, line 584 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXbXUkbNui7r5ArIl4:-KhXbXUkbNui7r5ArIl5:bn7vmis) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ba43f983ca7781cef587f4a0bb65541da40315a/pkg/util/log/clog.go#L584)):*
> ```Go
> 
> 	// Level flag for output to stderr. Handled atomically.
> 	stderrThreshold Severity // The -alsologtostderr flag.
> ```

is now the time to rename "alsologtostderr" to something more naturally implying a threshold?

---

*[pkg/util/log/clog_test.go, line 652 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZqE5wTlEWBBTKy4X:-KhXZqE6IdUvbcakt-p6:bowt4n6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ba43f983ca7781cef587f4a0bb65541da40315a/pkg/util/log/clog_test.go#L652)):*
> ```Go
> 
> func TestFileSeverityFilter(t *testing.T) {
> 	// This test requires that the logs go to files. We must disable
> ```

something something @bramgruneir's other PR changes this? y'all should figure out what the correct way to go is.

---

*[pkg/util/log/clog_test.go, line 674 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZvTTrduc63n74UCb:-KhXZvTTrduc63n74UCc:b-yt046a) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ba43f983ca7781cef587f4a0bb65541da40315a/pkg/util/log/clog_test.go#L674)):*
> ```Go
> 	}
> 	if !strings.Contains(string(contents), "test2") {
> 		t.Fatalf("log does not contain error text\n%s", contents)
> ```

seems like this could be Errorf, ditto below.

---

*[pkg/util/log/file.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZBNQzM8qcUr2_qSK:-KhXZBNQzM8qcUr2_qSL:bc8ujx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/log/file.go#L43)):*
> ```Go
> 
> // LogFileMaxSize is the maximum size of a log file in bytes.
> var LogFileMaxSize int64 = 1024 * 1024 * 10
> ```

nit: `10 << 20 // 10 MiB`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 2 of 7 files at r2, 2 of 5 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/util/humanizeutil/humanize.go, line 84 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXYkep5aVCeuS4XdOH:-KhXeLLIjydJE85MKsqD:b9fp32j) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/humanizeutil/humanize.go#L84)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this doesn't seem to have anything to do with what the commit message claims. why was it necessary? what's the race?
</blockquote></details>

the log gcDaemon() is spawned in the init() method to run asynchronously. It uses the setting. This is racy with cmdline argument handling.

---

*[pkg/util/log/clog.go, line 902 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXYzPy2saFi2zv-AIj:-KhXetf_zKx5seg_t-IF:b-g50ymk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/log/clog.go#L902)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto - what's the race? here and below.
</blockquote></details>

asynchronous gcDaemon()

---

*[pkg/util/log/clog.go, line 89 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXbGz9mlDlvO4gazof:-KhXeUYssZWO5XY4MWOF:bjscee6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2a63d5851900063acfa188d957c7d6315323ade/pkg/util/log/clog.go#L89)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

was this just...completely wrong before?

why does it need to be atomic now?
</blockquote></details>

Yes it was wrong conceptually, but it happened to be OK in practice because the one `stderrThreshold` parameter happened to be the only one settable using its flag.Value interface.

Why atomic - this is what the set() method was originally doing, I am not changing anything about that.

---

*[pkg/util/log/clog.go, line 584 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXbXUkbNui7r5ArIl4:-KhXeqco-8MAza2-4JBW:buu1cbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ba43f983ca7781cef587f4a0bb65541da40315a/pkg/util/log/clog.go#L584)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is now the time to rename "alsologtostderr" to something more naturally implying a threshold?
</blockquote></details>

I agree. Do you have a suggestion?

---

*[pkg/util/log/clog_test.go, line 652 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZqE5wTlEWBBTKy4X:-KhXewZtBQGN5whCEbb6:b-o5hsfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ba43f983ca7781cef587f4a0bb65541da40315a/pkg/util/log/clog_test.go#L652)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

something something @bramgruneir's other PR changes this? y'all should figure out what the correct way to go is.
</blockquote></details>

Yes the two PRs are competing on this. Either of the first two that merges first will cause the other one to adapt.
As long as the other PR is not merged, this code remains correct though.

---

*[pkg/util/log/clog_test.go, line 674 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZvTTrduc63n74UCb:-KhXfPap3KhAc6Xy6ecb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ba43f983ca7781cef587f4a0bb65541da40315a/pkg/util/log/clog_test.go#L674)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

seems like this could be Errorf, ditto below.
</blockquote></details>

Done.

---

*[pkg/util/log/file.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZBNQzM8qcUr2_qSK:-KhXfQPU73Q_FFPuuQnY:bb43dg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/log/file.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: `10 << 20 // 10 MiB`
</blockquote></details>

This is evaluated at compile time, I find the original version easier on the eye.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/util/humanizeutil/humanize.go, line 84 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXYkep5aVCeuS4XdOH:-KhXh2Ce2hzsB2cNvnaQ:bjvgv98) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/humanizeutil/humanize.go#L84)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

the log gcDaemon() is spawned in the init() method to run asynchronously. It uses the setting. This is racy with cmdline argument handling.
</blockquote></details>

Comment! Also perhaps include it in the commit message.

---

*[pkg/util/log/clog.go, line 89 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXbGz9mlDlvO4gazof:-KhXhO7eSWk_GL6tIhtO:b207ffh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2a63d5851900063acfa188d957c7d6315323ade/pkg/util/log/clog.go#L89)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes it was wrong conceptually, but it happened to be OK in practice because the one `stderrThreshold` parameter happened to be the only one settable using its flag.Value interface.

Why atomic - this is what the set() method was originally doing, I am not changing anything about that.
</blockquote></details>

What do you mean? It wasn't using atomics before.

---

*[pkg/util/log/clog.go, line 584 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXbXUkbNui7r5ArIl4:-KhXhSWygfzE9VKbMGyO:b9vez6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ba43f983ca7781cef587f4a0bb65541da40315a/pkg/util/log/clog.go#L584)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I agree. Do you have a suggestion?
</blockquote></details>

I don't, but perhaps something similar to the name of the new flag you're adding (the actual name has already left my mental cache).

---

*[pkg/util/log/file.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZBNQzM8qcUr2_qSK:-KhXhGGE2TWf8MNpjxU0:b-hyrcza) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/log/file.go#L43)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This is evaluated at compile time, I find the original version easier on the eye.
</blockquote></details>

There's lots of precedent for the bit-shift style, though =/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

knz committed:
log: wait until cmdline args have been parsed to start the GC

Prior to this patch, the log file garbage collector would start
immediately upon process start-up, and thus start removing files based
on the default max sizes (e.g. 100MiB for the entire log directory).
This would cause unintended log file loss if the user intends the
maximum size to be larger than the default.

Also there was a race between the GC and the initialization of the log
directory, which would have caused the GC to fail removing files if
the command line parsing was occuring concurrently with a GC run.

This patch ensures the GC is only started after the command-line
parameters have been set.

knz committed:
log: remove --alsologtostderr; make --logtostderr decide the verbosity.

This patch follows up on 98c47de36bdbd9be1d80ac4dcb60db183a4d9f86 and
simplifies the logic further: now `--logtostderr` also accepts a severity
threshold for the messages that go to stderr. The default threshold
is dependent on the top-level command being invoked.

For documentation, let's clarify this as follows:

- now any logging message is a candidate to be printed on stderr,
  subject to the configured *severity threshold*, regardless of
  whether `--log-dir` is specified.

- the severity threshold can always be overridden via
  `--logtostderr=XXX`.

- The severity threshold gets a default value that depends on the
  command being run and other things:

  - for `start`:

    - `--log-dir` specified and non-empty, or there is at least one
      disk-based store. Then `--logtostderr` decides, either the
      specified value if one is specified, or INFO (everything) if the
      flag is set but no value is specified, or NONE (nothing) if the
      flag is not present or the value is "false".

    - `--log-dir` not specified or empty, and only in-memory stores:
      default threshold is INFO.

  - for other CLI commands:

    - `--log-dir` specified and non-empty: `--logtostderr` decides,
      either the specified value if one is specified, or WARNING if
      the flag is set but no value is specified, or NONE (nothing) if
      the flag is not present or the value is "false".

    - `--log-dir` not specified or empty: default threshold is
      WARNING.

knz commented:
Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/util/humanizeutil/humanize.go, line 84 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXYkep5aVCeuS4XdOH:-KhXlyUji0L825-bTgHR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/humanizeutil/humanize.go#L84)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Comment! Also perhaps include it in the commit message.
</blockquote></details>

Done.

---

*[pkg/util/log/clog.go, line 89 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXbGz9mlDlvO4gazof:-KhXkpQToOkaJARl-0EU:b-v4qc5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2a63d5851900063acfa188d957c7d6315323ade/pkg/util/log/clog.go#L89)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

What do you mean? It wasn't using atomics before.
</blockquote></details>

It was: the method `(*Severity) set(val Severity)` which was used by the original code uses `atomic.StoreInt32`. Anyway, I changed the code to use that method.

---

*[pkg/util/log/clog.go, line 584 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXbXUkbNui7r5ArIl4:-KhYLO395A1PrMGuAUoW:b-uz6nae) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ba43f983ca7781cef587f4a0bb65541da40315a/pkg/util/log/clog.go#L584)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I don't, but perhaps something similar to the name of the new flag you're adding (the actual name has already left my mental cache).
</blockquote></details>

Done. I simply repurposed the existing flag `--logtostderr`.

---

*[pkg/util/log/file.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZBNQzM8qcUr2_qSK:-KhXldR899lArlK0JpkI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/log/file.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

There's lots of precedent for the bit-shift style, though =/
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/util/log/file.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZBNQzM8qcUr2_qSK:-KhYLwI0hi_lBN2W5Yvy:b23uu36) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/log/file.go#L43)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think the reason why there's lots of precedent is because you keep pushing it on people :)
I also find it less readable and had written things like here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

tamird commented:
You should check the git blame before assigning ... 🕶️ blame.

On Apr 12, 2017 2:11 PM, "Andrei Matei" <notifications@github.com> wrote:

> Review status: 0 of 18 files reviewed at latest revision, 5 unresolved
> discussions.
> ------------------------------
>
> *pkg/util/log/file.go, line 43 at r2
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZBNQzM8qcUr2_qSK:-KhYLwI0hi_lBN2W5Yvy:b23uu36>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/2d47d2f972f0c7889c4867660079c55d883455c4/pkg/util/log/file.go#L43>):*
> *Previously, knz (kena) wrote…*
>
> Done.
>
> I think the reason why there's lots of precedent is because you keep
> pushing it on people :)
> I also find it less readable and had written things like here.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/14841>*
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14841#issuecomment-293662093>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPLByGf9ADICnusenpob4epyfHJTOks5rvRO4gaJpZM4M7OS3>
> .
>

andreimatei commented:
well I don't know who started it, but I've seen you push it on other people's code (so blame doesn't help). Anyway, this is an example where, in my opinion, consistency is not needed.

---

Review status: 0 of 18 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-:-KhYOM7rLsrMxMz3SpDv:bug612k)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
*[pkg/util/log/clog_test.go, line 652 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhXZqE5wTlEWBBTKy4X:-KhYVD5f5ukEV49fLAuT:bb1232l) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ba43f983ca7781cef587f4a0bb65541da40315a/pkg/util/log/clog_test.go#L652)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes the two PRs are competing on this. Either of the first two that merges first will cause the other one to adapt.
As long as the other PR is not merged, this code remains correct though.
</blockquote></details>

It's a race!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Reviewed 2 of 2 files at r1, 9 of 9 files at r9, 2 of 7 files at r10, 16 of 16 files at r16.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/cli/flags.go, line 254 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhYWWbdyBXkIdBrIHMg:-KhYWWbdyBXkIdBrIHMh:bwrvp7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/207e8bd5f7bf9348463ea0af4959939549e01197/pkg/cli/flags.go#L254)):*
> ```Go
> 	// If no value is specified for --logtostderr prepare for the
> 	// default in `setDefaultStderrVerbosity()`.
> 	pf.Lookup(logflags.LogToStderrName).NoOptDefVal = "unknown"
> ```

Perhaps instead of setting this manually, log.Serverity_UNKNOWN.String() ?

---

*[pkg/util/log/clog.go, line 573 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhYXONbDq4l5CiYG2OV:-KhYXONbDq4l5CiYG2OW:b-e4kljf) ([raw file](https://github.com/cockroachdb/cockroach/blob/207e8bd5f7bf9348463ea0af4959939549e01197/pkg/util/log/clog.go#L573)):*
> ```Go
> // user configures larger max sizes than the defaults.
> func StartGCDaemon() {
> 	go logging.gcDaemon()
> ```

It might be nice to find a way to be able to stop and restart the deamon instead of using an atomic as I did in #14831.

---

*[pkg/util/log/clog.go, line 719 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhYXqOvcbwevt3jnedX:-KhYXqOvcbwevt3jnedY:b-hygne4) ([raw file](https://github.com/cockroachdb/cockroach/blob/207e8bd5f7bf9348463ea0af4959939549e01197/pkg/util/log/clog.go#L719)):*
> ```Go
> 		l.outputToStderr(entry, stacks)
> 	}
> 	if logDir.isSet() && s >= l.fileThreshold.get() {
> ```

I've mentioned this before, but there must be a better way than having to do a bunch of individual atomic lookups in these functions.  Maybe not in this PR, but in a follow up.

---

*[pkg/util/log/doc.go, line 48 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhYZb1aWCWD2it7bJcP:-KhYZb1aWCWD2it7bJcQ:bf1gomd) ([raw file](https://github.com/cockroachdb/cockroach/blob/207e8bd5f7bf9348463ea0af4959939549e01197/pkg/util/log/doc.go#L48)):*
> ```Go
> //
> //	--logtostderr=INFO
> //		Logs are written to standard error as well as to files.
> ```

Add a comment on severity?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM. I haven't read the code in detail but :thumbsup: to the new command-line flags.

---

Reviewed 2 of 7 files at r10, 16 of 16 files at r16.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-:-Kh_0VENgQnEyObx4nAy:bdz6ng9)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 18 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/cli/flags.go, line 254 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhYWWbdyBXkIdBrIHMg:-KhanjdLl7wlFDXcfUL7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/207e8bd5f7bf9348463ea0af4959939549e01197/pkg/cli/flags.go#L254)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Perhaps instead of setting this manually, log.Serverity_UNKNOWN.String() ?
</blockquote></details>

Done.

---

*[pkg/util/log/clog.go, line 573 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhYXONbDq4l5CiYG2OV:-Khanle7ng82lEuUtlAn:bqjrgzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/207e8bd5f7bf9348463ea0af4959939549e01197/pkg/util/log/clog.go#L573)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

It might be nice to find a way to be able to stop and restart the deamon instead of using an atomic as I did in #14831.
</blockquote></details>

Usually people do this using a channel and a boolean flag through the channel. I'd still prefer to see that in a separate PR though.

---

*[pkg/util/log/clog.go, line 719 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhYXqOvcbwevt3jnedX:-KhanUQ3pVvgz6cU1K0k:b-kd1al2) ([raw file](https://github.com/cockroachdb/cockroach/blob/207e8bd5f7bf9348463ea0af4959939549e01197/pkg/util/log/clog.go#L719)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

I've mentioned this before, but there must be a better way than having to do a bunch of individual atomic lookups in these functions.  Maybe not in this PR, but in a follow up.
</blockquote></details>

I'd be glad to hear it. Feel free to make suggestions.

---

*[pkg/util/log/doc.go, line 48 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhYZb1aWCWD2it7bJcP:-KhanfridLTC-rU55vgT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/207e8bd5f7bf9348463ea0af4959939549e01197/pkg/util/log/doc.go#L48)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Add a comment on severity?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

knz commented:
@tamird can you please regenerate `embedded.go` in the last commit please. Thanks
(I can't find any machine here with a suitable combination of tool versions)

knz commented:
Never mind, I was able to re-generate embedded.go after all.

tamird commented:
Reviewed 2 of 8 files at r18, 21 of 21 files at r24.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/cli/cli_test.go, line 1158 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khbio4fpR0M2WQyKsmA:-Khbio4fpR0M2WQyKsmB:b9a9tuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/cli/cli_test.go#L1158)):*
> ```Go
>       --log-file-max-size bytes          maximum size of each log file (default 10 MiB)
>       --log-file-verbosity Severity      minimum verbosity of messages written to the log file (default INFO)
>       --logtostderr Severity[=DEFAULT]   logs at or above this threshold go to stderr (default NONE)
> ```

how come this defaults to NONE, while the one above defaults to INFO?

If there's no reason, we shouldn't change the protos; UNKNOWN is meant as a sentinel, not a real value that we use.

---

*[pkg/util/log/clog.go, line 590 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhbjwChOPCcbRiXIaWw:-KhbjwChOPCcbRiXIaWx:b-bykl3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/util/log/clog.go#L590)):*
> // The -logtostderr flag.

I'd be ok with removing this; you don't include it for `fileThreshold`.

---

*[pkg/util/log/log.proto, line 24 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khbk7C9y60gsRXyJQh8:-Khbk7C9y60gsRXyJQh9:bioovvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/util/log/log.proto#L24)):*
> ```protobuf
> 
> enum Severity {
>   DEFAULT = 0;
> ```

as mentioned elsewhere, I don't think this change is necessary or correct.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 1158 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khbio4fpR0M2WQyKsmA:-KhbuKE5Vgio7IbZ1nca:bf9o00k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/cli/cli_test.go#L1158)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

how come this defaults to NONE, while the one above defaults to INFO?

If there's no reason, we shouldn't change the protos; UNKNOWN is meant as a sentinel, not a real value that we use.
</blockquote></details>

There are two unrelated points here.
1) it defaults to NONE because logging to stderr is disabled by default. That's OK and what we want.
2) the thing on the left ("[=DEFAULT])") is not really the default; it's the value used by pflag if the flag `--logtostderr` is specified but not followed by an explicit severity. In this case, we want the value to depend on which command is used. Now, if I did not change the protobuf (in particular the string representation of value 0), the usage text generated by pflag would be: "`--logtostderr Severity[=UNKNOWN]   logs at or above this threshold go to stderr (default NONE)`" which I think is poor UX.

I can see three ways forward:
- change pflag to generate a custom usage line that does not mention "UNKNOWN" -- I do not know how to do that
- introduce a new severity value after NONE called "DEFAULT" and use that as sentinel here instead
- keep the protobuf change

What do you think is best?

---

*[pkg/util/log/clog.go, line 590 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhbjwChOPCcbRiXIaWw:-KhbvIr8ILk6KEqhKNGF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/util/log/clog.go#L590)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> // The -logtostderr flag.

I'd be ok with removing this; you don't include it for `fileThreshold`.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r25.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 1158 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khbio4fpR0M2WQyKsmA:-Khbvx7fJDIGYznClTbl:b-951u49) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/cli/cli_test.go#L1158)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

There are two unrelated points here.
1) it defaults to NONE because logging to stderr is disabled by default. That's OK and what we want.
2) the thing on the left ("[=DEFAULT])") is not really the default; it's the value used by pflag if the flag `--logtostderr` is specified but not followed by an explicit severity. In this case, we want the value to depend on which command is used. Now, if I did not change the protobuf (in particular the string representation of value 0), the usage text generated by pflag would be: "`--logtostderr Severity[=UNKNOWN]   logs at or above this threshold go to stderr (default NONE)`" which I think is poor UX.

I can see three ways forward:
- change pflag to generate a custom usage line that does not mention "UNKNOWN" -- I do not know how to do that
- introduce a new severity value after NONE called "DEFAULT" and use that as sentinel here instead
- keep the protobuf change

What do you think is best?
</blockquote></details>

Well, the value printed here comes from the flag's `NoOptDefVal` property; don't we want that to be `NONE`?

---

*[pkg/cli/flags.go, line 254 at r25](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhbwKojFTtoXXyjw6Uv:-KhbwKojFTtoXXyjw6Uw:b-n5hxbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/beae8f8d0fde0895267c666665096e06ff45de14/pkg/cli/flags.go#L254)):*
> ```Go
> 	// If no value is specified for --logtostderr prepare for the
> 	// default in `setDefaultStderrVerbosity()`.
> 	pf.Lookup(logflags.LogToStderrName).NoOptDefVal = log.Severity_DEFAULT.String()
> ```

how come this can't be `log.Severity_NONE.String()`? Also, the comment above this line doesn't make any sense to me, what does this have to do with preparing for `setDefaultStderrVerbosity`?

As I understand it, this flag is created in the log package, but it's a simple go "flag" there, and here we want to set the extra NoOptDefVal property which is only present in pflag. I think you can just remove this comment and be done with it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/cli/flags.go, line 254 at r25](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhbwKojFTtoXXyjw6Uv:-KhbxIVA7dY7BzGsfurH:b-jcfy9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/beae8f8d0fde0895267c666665096e06ff45de14/pkg/cli/flags.go#L254)):*
> how come this can't be log.Severity_NONE.String()?

Ah, sorry, this should be `log.Severity_INFO.String()`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 1158 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khbio4fpR0M2WQyKsmA:-Khbxlxuw13hhu3XJcyn:b-qyommd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/cli/cli_test.go#L1158)):*
> In this case, we want the value to depend on which command is used.

Oh, I think I understand now. You just want the word "default"? I think you can just set NoOptDefVal to "DEFAULT" and get the desired behaviour, since we never actually look at this implicit default value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 1158 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khbio4fpR0M2WQyKsmA:-Khc8ev5pHfmWAkwwM_J:bxia03j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/cli/cli_test.go#L1158)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> In this case, we want the value to depend on which command is used.

Oh, I think I understand now. You just want the word "default"? I think you can just set NoOptDefVal to "DEFAULT" and get the desired behaviour, since we never actually look at this implicit default value.
</blockquote></details>

That's not correct, unfortunately. By the time `PersistentPreRunE` runs and calls `setDefaultStderrVerbosity` where this is checked, `pflag` has already called the `Set()` method on the value. That will have tried to convert the string to a severity, and fail.

Since I sense you're reticent to rename UNKNOWN, I'll go for an external proto value higher than NONE to represent DEFAULT.

---

*[pkg/cli/flags.go, line 254 at r25](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhbwKojFTtoXXyjw6Uv:-Khc8zDvcRmolMQaY2K2:b-jnhbq1) ([raw file](https://github.com/cockroachdb/cockroach/blob/beae8f8d0fde0895267c666665096e06ff45de14/pkg/cli/flags.go#L254)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> how come this can't be log.Severity_NONE.String()?

Ah, sorry, this should be `log.Severity_INFO.String()`.
</blockquote></details>

See my explanation above: by the time `setDefaultStderrVerbosity()` pflag will have called the `Set()` method and tried to convert the value.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 1158 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khbio4fpR0M2WQyKsmA:-KhbxU0pqkpjgyx9WEwr:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/cli/cli_test.go#L1158)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

That's not correct, unfortunately. By the time `PersistentPreRunE` runs and calls `setDefaultStderrVerbosity` where this is checked, `pflag` has already called the `Set()` method on the value. That will have tried to convert the string to a severity, and fail.

Since I sense you're reticent to rename UNKNOWN, I'll go for an external proto value higher than NONE to represent DEFAULT.
</blockquote></details>

:+1:

---

*[pkg/cli/flags.go, line 254 at r25](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhbwKojFTtoXXyjw6Uv:-KhcGhMh8CVpY06FElOW:b-6ryvw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/beae8f8d0fde0895267c666665096e06ff45de14/pkg/cli/flags.go#L254)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

See my explanation above: by the time `setDefaultStderrVerbosity()` pflag will have called the `Set()` method and tried to convert the value.
</blockquote></details>

I see. Still, the comment above seems misleading at best (something to keep in mind when you make the change you refer to above).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/cli/cli_test.go, line 1158 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khbio4fpR0M2WQyKsmA:-KhcXG7_Cq_-uHSLXv4c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/cli/cli_test.go#L1158)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

:+1:
</blockquote></details>

Done.

---

*[pkg/cli/flags.go, line 254 at r25](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhbwKojFTtoXXyjw6Uv:-KhcXszWN007pZhdLNLD:b-mb71kk) ([raw file](https://github.com/cockroachdb/cockroach/blob/beae8f8d0fde0895267c666665096e06ff45de14/pkg/cli/flags.go#L254)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I see. Still, the comment above seems misleading at best (something to keep in mind when you make the change you refer to above).
</blockquote></details>

I reworded the comment to (attempt to) clarify.

---

*[pkg/util/log/log.proto, line 24 at r24](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khbk7C9y60gsRXyJQh8:-KhcXvuhjzWX0CFkbksv:bp3aow6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b34bfb55dd11f4c4d3697c9397f73982380e054a/pkg/util/log/log.proto#L24)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

as mentioned elsewhere, I don't think this change is necessary or correct.
</blockquote></details>

Reverted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 8 files at r27, 21 of 21 files at r33.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/cli/flags.go, line 254 at r25](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhbwKojFTtoXXyjw6Uv:-Khc_SmUtlp2OarakPwP:b-sbz8ny) ([raw file](https://github.com/cockroachdb/cockroach/blob/beae8f8d0fde0895267c666665096e06ff45de14/pkg/cli/flags.go#L254)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I reworded the comment to (attempt to) clarify.
</blockquote></details>

I don't see any change?

---

*[pkg/cli/flags.go, line 440 at r33](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khc_XJp-Fb0wTHB_sn-:-Khc_XJp-Fb0wTHB_sn0:bim7yw1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8463eb896a4e6e4afde2a85437c29637eb1bee/pkg/cli/flags.go#L440)):*
> ```Go
> 	vf := pf.Lookup(logflags.LogToStderrName)
> 
> 	// if `--logtostderr` was not specified but no log directory was
> ```

s/but/and/ while you're here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-:-Khc_shjh8m1ale9gGY6:bnfp4nl)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 22 of 23 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/cli/flags.go, line 254 at r25](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-KhbwKojFTtoXXyjw6Uv:-KhcbCmJa2fuG_elwHWR:b-duu5u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/beae8f8d0fde0895267c666665096e06ff45de14/pkg/cli/flags.go#L254)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I don't see any change?
</blockquote></details>

Forgot to commit. Done.

---

*[pkg/cli/flags.go, line 440 at r33](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khc_XJp-Fb0wTHB_sn-:-KhcbFbXPkL6eCrN3aJF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e8463eb896a4e6e4afde2a85437c29637eb1bee/pkg/cli/flags.go#L440)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/but/and/ while you're here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

knz commented:
yay it's green now

tamird commented:
Reviewed 1 of 1 files at r34.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/cli/flags.go, line 460 at r34](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841#-Khck3acI38zOq30e1eS:-Khck3acI38zOq30e1eT:btyt8bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fdc51ee4340e109fc2995e29ebae6a41f97d153/pkg/cli/flags.go#L460)):*
> ```Go
> 	// If `--logtostderr` was specified without explicit verbosity,
> 	// set to the level that is considered default by the specific command.
> 	if vf.Value.String() == log.Severity_DEFAULT.String() {
> ```

nit: this can be coalesced into the above condition, but it's fine as is, too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14841)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for all the support and good suggestions!

knz commented:
oops, we crossed fire

knz commented:
I'll follow up with as eparate commit
